### PR TITLE
User UI camera control

### LINKED ISSUE #5242: UI camera flexibility regression
## Bevy version

bf1ca81779

## What you did

In bevy 0.7, it was possible to manipulate individually ui cameras such as [in this example](https://github.com/nicopap/ui-navigation/blob/30c828a465f9d4c440d0ce6e97051a5f7fafa425/examples/off_screen_focusables.rs).

## What went wrong

With changes made to the way ui is handled in #4745, such design is impossible to execute. Manipulating UI cameras can be useful for fancy game effects, and it would be a shame to lose this flexibility.

### PR DESCRIPTION
**Pull request by nicopap**
# Objective

Enable user manipulation of UI camera.
* Fixes #5242
* Supersedes #5243
* Depends on (the **two first** commits of this PR) #5414 

### Add user control of UI cameras

Add fields to UiCameraConfig to control the UI camera scale and
position.

Fixes #5242. It is again possible to manipulate the ui camera.

An alternative design was considered, where instead of having a
component that controls all of the UI camera settings, the component
would only hold an Entity referencing another camera.

That design was abandoned in favor of the current one because the
viewport is tightly bound to the "actual" camera the UI camera is
attached to. So it would be awkward to maintain independently two
different cameras.

---

## Changelog

- Add `position`, `scale` and `window_origin` fields to `UiCameraConfig` to enable
  manipulating the position of the attached UI camera.
- Add UI Camera control example
- Fix UI focus not working properly with moving cameras

## Migration Guide

- Instead of spawning an individual ui camera with `UiCameraBundle`
  and manipulating its `OrthographicProjection` and `GlobalTransform`,
  add a `UiCameraConfig` component to your viewport camera, and
  use its field to change the ui camera position
- Use the `UiCameraProjection` component to access the UI camera's projection.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Sibling to #5114 and #4007.

**Comment by alice-i-cecile** [comment]
@IceSentry @superdump, I think this is a serious enough breakage that we should try to get this reviewed and merged for 0.8

File: crates/bevy_ui/src/render/mod.rs [hunk 1]
```diff
@@ -356,8 +359,10 @@ const QUAD_VERTEX_POSITIONS: [Vec3; 4] = [
 
 const QUAD_INDICES: [usize; 6] = [0, 2, 3, 0, 1, 2];
 
+// Ui nodes are batches per image and per layer
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/render/mod.rs:362]:
```suggestion
/// UI nodes are batched per image and per layer
```

File: examples/window/multiple_windows.rs [hunk 2]
```diff
@@ -50,12 +55,41 @@ fn setup(
...
+        .spawn_bundle(Camera3dBundle {
+            transform: Transform::from_xyz(6.0, 0.0, 0.0).looking_at(Vec3::ZERO, Vec3::Y),
+            camera: Camera {
+                target: RenderTarget::Window(window_id),
+                ..default()
+            },
             ..default()
-        },
+        })
+        .insert(UiCameraConfig {
```
Comment (alice-i-cecile) [inline examples/window/multiple_windows.rs:67]: This part of the example is unclear, and needs more comments.

We can also probably shorten it now that #5343 has merged!

File: examples/ui/ui.rs [hunk 3]
```diff
@@ -309,6 +316,31 @@ struct ScrollingList {
     position: f32,
 }
 
+fn change_ui_camera(
```
Comment (alice-i-cecile) [inline examples/ui/ui.rs:319]: IMO we should have this in its own example. I don't like how big this example is to begin with; let's not make it worse.

**Comment by nicopap** [comment]
@IceSentry said:
> I feel like this should be split in 2 PR. I like the render layer idea, but I don't like the config part because of the pub projection.

Indeed. In insight, it's quite obvious those are two different features.

The `RenderLayer` addition however, leaves #5242 unsolved. I agree the `pub projection` is questionable, and I've spent a some time (let's be honest, few minutes) thinking about it.

The thing is, you _need_ access to the UI projection for a lot of reasons (navigation is one of them).

Here is a different proposition: have a `UiProjection` component which public API provide read-only access to the projection, have a system to update it based on the associated camera and the config. I'm still very annoyed by that solution, because you then have to add systems relying on the projection to a specific stage and make sure they run `.after` the ui projection update system.

But the one I provided here suffers from the same issue, so maybe it's OK?

**Comment by nicopap** [comment]
> Mhmm. I think we split this, merge the `RenderLayers` part ASAP, and then chew on the config in this PR.

Can do, ready tomorrow.

**Comment by nicopap** [comment]
I created a separate PR for the RenderLayers, see #5414. Note that this PR still contains all the commits from #5414, so I expect it to be reviewed later on.

**Comment by nicopap** [comment]
This PR is not fully addressed by #5892, but I think it should be the direction forward, thus superseding this PR.

**Pull request closed without merging**